clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -1032,25 +1036,89 @@ raw2trace_t::process_syscall_pt(raw2trace_thread_data_t *tdata, uint64_t syscall
+    trace_entry_t entries_with_encodings[WRITE_BUFFER_SIZE];
+    trace_entry_t *buf = entries_with_encodings;
     for (const auto &entry : entries) {
-        if (type_is_instr(static_cast<trace_type_t>(entry.type)))
+        if (type_is_instr(static_cast<trace_type_t>(entry.type))) {
+            if (buf != entries_with_encodings) {
+                if (!write(tdata, entries_with_encodings, buf, &saved_decode_pc, 1)) {
+                    return false;
+                }
+                buf = entries_with_encodings;
+            }
+            app_pc instr_pc = reinterpret_cast<app_pc>(entry.addr);
             accumulate_to_statistic(tdata, RAW2TRACE_STAT_KERNEL_INSTR_COUNT, 1);
+            if (tdata->syscall_pc_to_decode_pc_.find(instr_pc) ==
+                tdata->syscall_pc_to_decode_pc_.end()) {
+                tdata->error =
+                    "Unknown pc after ir2trace: did ir2trace insert new instr?";
+                return false;
+            }
+            saved_decode_pc = tdata->syscall_pc_to_decode_pc_[instr_pc].first;
+            if (!append_encoding(tdata, saved_decode_pc, entry.size, buf,
+                                 entries_with_encodings))
+                return false;
+        }
+        *buf = entry;
+        ++buf;
     }
-
-    if (!tdata->out_file->write(reinterpret_cast<const char *>(entries.data()),
-                                sizeof(trace_entry_t) * entries.size())) {
-        tdata->error = "Failed to write to output file";
-        return false;
+    if (buf != entries_with_encodings) {
+        if (!write(tdata, entries_with_encodings, buf, &saved_decode_pc, 1)) {
+            return false;
+        }
     }
+    return true;
+}
 
+bool
+raw2trace_t::track_syscall_pt_encodings(raw2trace_thread_data_t *tdata,
+                                        instrlist_t *ilist)
+{
+    instr_t *instr = instrlist_first(ilist);
+    while (instr != NULL) {
+        app_pc orig_pc = instr_get_app_pc(instr);
+        int instr_len = instr_length(dcontext_, instr);
+        if (tdata->syscall_instr_encodings_.empty() ||
+            tdata->syscall_next_encoding_offset_ + instr_len >=
+                SYSCALL_PT_ENCODING_BUF_SIZE) {
+            // We allocate new memory to store kernel instruction encodings in
+            // increments of SYSCALL_PT_ENCODING_BUF_SIZE. Note that we cannot
+            // treat this like a cache and clear previously stored encodings
+            // because any re-use in the decode address will confuse our write()
+            // logic (remember that we key encodings using the decode pc).
+            tdata->syscall_instr_encodings_.emplace_back(
+                new uint8_t[SYSCALL_PT_ENCODING_BUF_SIZE]);
+            tdata->syscall_next_encoding_offset_ = 0;
+        }
+        app_pc encode_pc =
+            &tdata->syscall_instr_encodings_.back()[tdata->syscall_next_encoding_offset_];
+        byte *ret = instr_encode_to_copy(dcontext_, instr, encode_pc, orig_pc);
```

Comment:
We already have a copy of the original encoding bits, so I don't think we need to re-encode somewhere else.  See comment above where I'm confused as to why we need multiple layers of copies.

Reply:
Done. Removed the encode, reusing the bits from PT directly.